core/services/ocr2/plugins/vault/plugin.go

```diff
@@ -2,22 +2,756 @@ package vault
+	protoBytes, err := proto.MarshalOptions{Deterministic: true}.Marshal(msg)
+	if err != nil {
+		return "", fmt.Errorf("could not generate sha for proto message: failed to marshal proto: %w", err)
+	}
+
+	return fmt.Sprintf("%x", sha256.Sum256(protoBytes)), nil
+}
+
+func shaForObservation(o *vault.Observation) (string, error) {
+	switch o.RequestType {
+	case vault.RequestType_GET_SECRETS:
+		cloned := proto.CloneOf(o)
+		for _, r := range cloned.GetGetSecretsResponse().Responses {
+			if r.GetData() != nil {
+				// Exclude the encrypted shares from the sha, as these need to be aggregated later.
+				r.GetData().EncryptedDecryptionKeyShares = nil
+			}
+		}
+
+		return shaForProto(cloned)
+	default:
+		return shaForProto(o)
+	}
+}
+
+func validateObservation(o *vault.Observation) error {
+	if o.Id == "" {
+		return fmt.Errorf("observation id cannot be empty")
+	}
+
+	switch o.RequestType {
+	case vault.RequestType_GET_SECRETS:
+		if o.GetGetSecretsRequest() == nil || o.GetGetSecretsResponse() == nil {
+			return errors.New("GetSecrets observation must have both request and response")
+		}
+
+		if len(o.GetGetSecretsRequest().Requests) != len(o.GetGetSecretsResponse().Responses) {
+			return errors.New("GetSecrets request and response must have the same number of items")
+		}
+	case vault.RequestType_CREATE_SECRETS:
+		if o.GetCreateSecretsRequest() == nil || o.GetCreateSecretsResponse() == nil {
+			return errors.New("CreateSecrets observation must have both request and response")
+		}
+
+		if len(o.GetCreateSecretsRequest().EncryptedSecrets) != len(o.GetCreateSecretsResponse().Responses) {
+			return errors.New("CreateSecrets request and response must have the same number of items")
+		}
+
+		// We disallow duplicate create requests within a single batch request.
+		// This prevents users from clobbering their own writes.
+		idSet := map[string]bool{}
+		for _, r := range o.GetCreateSecretsRequest().EncryptedSecrets {
+			_, ok := idSet[keyFor(r.Id)]
+			if !ok {
+				idSet[keyFor(r.Id)] = true
+			} else {
+				return fmt.Errorf("CreateSecrets requests cannot contain duplicate request for a given secret identifier: %s", r.Id)
+			}
+		}
+	default:
+		return errors.New("invalid observation type: " + o.RequestType.String())
+	}
+
+	return nil
+}
+
+func (r *ReportingPlugin) StateTransition(ctx context.Context, seqNr uint64, aq types.AttributedQuery, aos []types.AttributedObservation, keyValueReadWriter ocr3_1types.KeyValueReadWriter, blobFetcher ocr3_1types.BlobFetcher) (ocr3_1types.ReportsPlusPrecursor, error) {
+	store := newWriteStore(keyValueReadWriter)
+
+	obsMap := map[string][]*vault.Observation{}
+	for _, ao := range aos {
+		obs := &vault.Observations{}
+		if err := proto.Unmarshal([]byte(ao.Observation), obs); err != nil {
+			r.lggr.Errorw("failed to unmarshal observations", "error", err, "observation", ao.Observation)
+			continue
+		}
+
+		for _, o := range obs.Observations {
+			err := validateObservation(o)
```

Comment:
Why not just use the ValidateObservations OCR phase for this?

Reply:
Good point :)